dna/codon.go

```diff
@@ -1,264 +1,245 @@
 	}
 }
 
-//NonSynonymous compares two Codon structs and returns true if they encode different AminoAcids.
-func NonSynonymous(c1 *Codon, c2 *Codon) bool {
-	return TranslateCodon(c1) != TranslateCodon(c2)
+// NonSynonymous compares two Codons and returns true if they encode different AminoAcids.
+func NonSynonymous(c1 Codon, c2 Codon) bool {
+	p1, _ := TranslateCodon(c1)
+	p2, _ := TranslateCodon(c2)
+	return p1 != p2
 }
 
-//Synonymous compares two codons and returns true if the codones have non-identical sequences that code for the same amino acid.
-func Synonymous(c1 *Codon, c2 *Codon) bool {
-	return TranslateCodon(c1) == TranslateCodon(c2) && !IsEqual(c1, c2)
+// Synonymous compares two codons and returns true if the codons code for the same amino acid.
+func Synonymous(c1 Codon, c2 Codon) bool {
+	p1, _ := TranslateCodon(c1)
+	p2, _ := TranslateCodon(c2)
+	return p1 == p2
 }
 
-//IsEqual compares two Codons and returns true if tthe underlying sequences are identical.
-func IsEqual(c1 *Codon, c2 *Codon) bool {
-	return BasesToString(c1.Seq) == BasesToString(c2.Seq)
+// IsEqual compares two Codons and returns true if the underlying sequences are identical.
+func IsEqual(c1 Codon, c2 Codon) bool {
+	return c1 == c2
 }
 
-//TranslateSeq takes a sequence of DNA bases and translates it into a slice of Amino acids.
-func TranslateSeq(b []Base) []AminoAcid {
-	var answer []AminoAcid
-	codons := BasesToCodons(b)
-	for i := 0; i < len(codons); i++ {
-		answer = append(answer, TranslateCodon(codons[i]))
+// TranslateSeq takes a sequence of DNA bases and translates it into a slice of Amino acids.
+func TranslateSeq(b []Base) ([]AminoAcid, error) {
+	answer := make([]AminoAcid, len(b)/3)
+	codons, err := BasesToCodons(b)
+	for i := range codons {
+		answer[i], err = TranslateCodon(codons[i])
 	}
-	return answer
+	return answer, err
 }
 
-//PolypeptideToShortString converts a slice of amino acid into a string of one character amino acid symbols.
+// PolypeptideToShortString converts a slice of amino acid into a string of one character amino acid symbols.
 func PolypeptideToShortString(a []AminoAcid) string {
-	var answer string
-	for i := 0; i < len(a); i++ {
-		answer = answer + AminoAcidToShortString(a[i])
+	var s strings.Builder
+	s.Grow(len(a))
+	for i := range a {
+		s.WriteByte(byte(a[i]))
 	}
-	return answer
+	return s.String()
 }
 
-//PolypeptideToString converts a slice of AminoAcids into a string of three character amino acid symbols.
-func PolypeptideToString(a []AminoAcid) string {
-	var answer string
-	for i := 0; i < len(a); i++ {
-		answer = answer + AminoAcidToString(a[i])
+// PolypeptideToString converts a slice of AminoAcids into a string of three character amino acid symbols.
+func PolypeptideToString(a []AminoAcid) (string, error) {
+	var s strings.Builder
+	s.Grow(len(a) * 3)
+	var str string
+	var err, finalErr error
+	for i := range a {
+		str, err = AminoAcidToString(a[i])
+		if err != nil {
+			finalErr = err
```

Comment:
This relates to a previous comment where we would need to think if the vision is to process data until the first error and then return the "error-free" part, or if we want to keep going and do the best with what we have.  One option would be to return right here.